### ISSUE TITLE
Split results in empty rows; empty rows are interpreted as separate records

### ISSUE THREAD

**Issue by strarsis** [OpenRefine/OpenRefine#3711]
### To Reproduce
1. Ensure a dataset that contains at least two records, each with multiple rows.
2. Switch to record mode.
3. Insert some delimited text into the cell of a row just before another record.
Use something that splits to empty cells (e.g. `test1;test2;;;;`)
4. Use the split cell feature with the right delimiter character.
Notice that at last one new row of empty lines has been inserted.
Notice that the row or rows don't belong to the record above but are counted as separate records.

### Current Results
Split cell feature can result in empty rows below, being in record mode, that are counted as separate records.

### Expected Behavior
Rows created by the split cell feature, being in record mode, resulting in rows that still belong to the originating record above.

### Versions
 - Operating System: Windows 10 x64
 - Browser Version: Chrome
 - OpenRefine: OpenRefine Windows Bundle

### Additional context
Besides from this issue, how can these empty rows/records be cleaned up? Is there a deletion feature - or is an external edit necessary?

**Comment by wetneb** [comment]
That's one of the many weaknesses of the records mode. I am not sure what motivated the choice of treating empty rows as new records. This is something we could change quite easily in the new architecture (4.x) but I would refrain from changing this in 3.x since it could be seen as a sneaky breaking change.

**Comment by strarsis** [comment]
@wetneb: Any quick workaround for getting rid of all the empty rows?

**Comment by ostephens** [comment]
@strarsis you can do a custom facet to find completely empty rows. E.g. custom text facet with GREL:

filter(row.columnNames,cn,isNonBlank(cells[cn].value)).length()==0

Will give "true" where all cells in the row are empty

**Comment by strarsis** [comment]
@ostephens: That's great! Is there also the possibility to remove all these matching, blank records?

**Comment by ostephens** [comment]
Do you mean just removing the blank rows? Once you have filtered to the blank rows in the "All" column header there is the option "Edit Rows" -> "Remove matching rows"

**Comment by thadguidry** [comment]
You might also eliminate the creation of the blank rows during the import
process itself by unchecking the default "Store blank rows".

Thad
https://www.linkedin.com/in/thadguidry/
https://calendly.com/thadguidry/

**Comment by strarsis** [comment]
@thadguidry: Thanks, though these extra, blank rows were created after the import by OpenRefine (cell split issue).
But this option could still be useful if the blank rows should be cleaned up using re-import.

**Comment by ostephens** [comment]
@strarsis going back to the original behaviour that you raised, in the issue you say the expected behaviour is:
>Rows created by the split cell feature, being in record mode, resulting in rows that still belong to the originating record above.

I'm trying to understand this better. Are you able to give an example where treating these blank lines as part of the original record gives a better outcome in some way than treating them as new records?

From your comments, I'm wondering if the underlying requirement is simply to never create blank rows as  part of a multi-value cell split?

**Comment by strarsis** [comment]
@ostephens: Yes, not even new empty rows are inserted, but they also become new records.

**Comment by ostephens** [comment]
@strarsis yes I understand - but I don't understand what difference that makes? Why would it be better, from your perspective, for them to appear as empty rows in the existing record?

**Comment by wetneb** [comment]
> Why would it be better, from your perspective, for them to appear as empty rows in the existing record?

For me, it would be better for the sake of being able to represent the original data more faithfully. Perhaps these empty values are meaningful!
In general, I would expect that splitting multi-valued cells and then joining them back should give the original table. If the splitting operation automatically ignores the empty values, then the two operations are no longer inverses of each other.

**Comment by strarsis** [comment]
Well, the moment where the ordering is permanently changed the connection of those empty-rowed records is lost irreversibly and the empty rows can't be restored anymore.

**Comment by wetneb** [comment]
I plan to change the definition of records in 4.0 so that empty lines are not records of their own. This is motivated by a new reason: if the first column is left untouched by a row-wise operation, which only makes changes to other columns, then we should be able to expect that the records boundaries remain the same. If that is the case, then this lets us preserve pagination in more cases, avoid recomputing record boundaries (which saves time), and provide a better analysis of data flows in a series of operations.
